helper/lora-packet-tracker.cc

```diff
@@ -137,9 +137,35 @@ LoraPacketTracker::PacketReceptionCallback (Ptr<Packet const> packet, uint32_t g
                                  << " was successfully received at gateway "
                                  << gwId);
 
-      std::map<Ptr<Packet const>, PacketStatus>::iterator it = m_packetTracker.find (packet);
-      (*it).second.outcomes.insert (std::pair<int, enum PhyPacketOutcome> (gwId,
-                                                                           RECEIVED));
+
+      if(m_packetTracker.count(packet) == 1) //first reception
+        {
+          std::map<Ptr<Packet const>, PacketStatus>::iterator it = m_packetTracker.find (packet);
+          (*it).second.outcomes.insert (std::pair<int, enum PhyPacketOutcome> (gwId,
+                                                                           RECEIVED)); 
+        }
+      else
+        {
+          std::pair <std::multimap<Ptr<Packet const>, PacketStatus>::iterator, std::multimap<Ptr<Packet const>, PacketStatus>::iterator> ret;
+          ret = m_packetTracker.equal_range(packet); // find all instances of received packet
+          
+          int i = 1;
+
+          //loop through all instances and find the one which doesn't yet have an outcome at this gateway
+          for(std::multimap<Ptr<Packet const>, PacketStatus>::iterator it=ret.first; it != ret.second; ++it)
+            {
+                    
+              if((*it).second.outcomes.find(gwId) == (*it).second.outcomes.end()) 
+              {
+                NS_LOG_DEBUG("This is copy  " << i <<" of packet "<< packet);                
+                NS_LOG_DEBUG("This packet was sent at " << (*it).second.sendTime);
+                NS_LOG_DEBUG("It was not yet received by GW " << gwId << " logging it now as RECEIVED.");
+                (*it).second.outcomes.insert(std::pair<int, enum PhyPacketOutcome> (gwId,
+                                                                           RECEIVED)); 
+              }
+              i++;           
+            }
+        }
```

Comment:
This code is repeated throughout the file for each callback - is it possible to put it inside a function that is called by each callback?

Reply:
Moved to function.